File: backend/theourgia/core/divination/iching/engine.py
"""Deterministic I Ching cast engine.

A *cast* produces six lines, bottom-up. Each line is one of:

* ``old_yin`` (value 6) — a yin line that **changes** to yang.
* ``young_yang`` (value 7) — a static yang line.
* ``young_yin`` (value 8) — a static yin line.
* ``old_yang`` (value 9) — a yang line that **changes** to yin.

Two casting methods, each with its canonical probability distribution:

**Three-coin method.** Three coins per line, each "heads" = 3, each
"tails" = 2 (or vice versa; the distribution is symmetric). The sum
is one of {6, 7, 8, 9} with probabilities {1/8, 3/8, 3/8, 1/8}.

**Yarrow stalk method.** The traditional 50-stalk division produces
a different distribution: {6, 7, 8, 9} with probabilities
{1/16, 5/16, 7/16, 3/16}. Static yin is the most common outcome; the
two changing lines together appear less often than under coins,
reflecting the tradition that "the yarrow is the harder oracle to
move."

The engine simulates the *outcome distribution* — it doesn't
literally manipulate 50 stalks — using the seeded RNG. Same seed +
same method = same six lines.

The resulting six lines determine the **primary hexagram** (read as
the six lines exactly drawn) and, if any lines are changing, the
**transformation hexagram** (the six lines after flipping every
changing line). Both numbers are returned. Some readings have only a
primary; readings with all-static lines have no transformation.
"""

from __future__ import annotations

import enum
import hashlib
import random
from dataclasses import dataclass
from typing import Final

__all__ = [
    "CastMethod",
    "CastResult",
    "LineKind",
    "cast_three_coins",
    "cast_yarrow_stalks",
    "hexagram_for_lines",
    "iching_cast",
    "lines_for_hexagram",
]


class CastMethod(str, enum.Enum):
    """Which generation method to use for a cast."""

    THREE_COINS = "three_coins"
    YARROW_STALKS = "yarrow_stalks"


class LineKind(str, enum.Enum):
    """One of the four line outcomes."""

    OLD_YIN = "old_yin"  # 6 — changing yin → yang
    YOUNG_YANG = "young_yang"  # 7 — static yang
    YOUNG_YIN = "young_yin"  # 8 — static yin
    OLD_YANG = "old_yang"  # 9 — changing yang → yin

    @property
    def is_yang(self) -> bool:
        return self in {LineKind.YOUNG_YANG, LineKind.OLD_YANG}

    @property
    def is_changing(self) -> bool:
        return self in {LineKind.OLD_YIN, LineKind.OLD_YANG}

    @property
    def value_number(self) -> int:
        return {
            LineKind.OLD_YIN: 6,
            LineKind.YOUNG_YANG: 7,
            LineKind.YOUNG_YIN: 8,
            LineKind.OLD_YANG: 9,
        }[self]


@dataclass(frozen=True, slots=True)
class CastResult:
    """The outcome of a single cast."""

    lines: tuple[LineKind, LineKind, LineKind, LineKind, LineKind, LineKind]
    """Bottom-up — ``lines[0]`` is line 1 (bottom), ``lines[5]`` is line 6 (top)."""

    primary_hexagram: int
    """King Wen number (1-64) of the hexagram exactly drawn."""

    transformation_hexagram: int | None
    """King Wen number of the hexagram after changing lines flip,
    or ``None`` if no lines changed."""

    changing_lines: tuple[int, ...]
    """Indices (1-based, bottom-up) of changing lines. Empty when no
    transformation; convention puts these lines in numerical order."""

    method: CastMethod
    """Which generation method produced this cast."""


# ───── Probability distributions ──────────────────────────────────────


# Each entry is (cumulative_threshold, line_kind). The seeded RNG draws
# a float in [0, 1) and we walk this list to find the first threshold
# the draw is less than.
_COIN_DISTRIBUTION: Final[tuple[tuple[float, LineKind], ...]] = (
    (1 / 8, LineKind.OLD_YIN),         # P=1/8
    (1 / 8 + 3 / 8, LineKind.YOUNG_YANG),  # +3/8 = 1/2
    (1 / 8 + 3 / 8 + 3 / 8, LineKind.YOUNG_YIN),  # +3/8 = 7/8
    (1.0, LineKind.OLD_YANG),          # +1/8 = 1
)

_YARROW_DISTRIBUTION: Final[tuple[tuple[float, LineKind], ...]] = (
    (1 / 16, LineKind.OLD_YIN),                # P=1/16
    (1 / 16 + 5 / 16, LineKind.YOUNG_YANG),    # +5/16 = 6/16
    (1 / 16 + 5 / 16 + 7 / 16, LineKind.YOUNG_YIN),  # +7/16 = 13/16
    (1.0, LineKind.OLD_YANG),                  # +3/16 = 1
)


def _draw_line(rng: random.Random, distribution: tuple[tuple[float, LineKind], ...]) -> LineKind:
    """Walk the cumulative table; return the line whose interval the
    draw falls in."""
    x = rng.random()
    for threshold, kind in distribution:
        if x < threshold:
            return kind
    return distribution[-1][1]  # defensive; shouldn't reach


def _seeded_random(seed: str) -> random.Random:
    """Same primitive as the Tarot engine — SHA-256 → 64-bit int → RNG."""
    digest = hashlib.sha256(seed.encode("utf-8")).digest()
    seed_int = int.from_bytes(digest[:8], "big", signed=False)
    return random.Random(seed_int)


# ───── Cast primitives ────────────────────────────────────────────────


def cast_three_coins(seed: str) -> tuple[LineKind, ...]:
    """Six-line cast using the three-coin probability table."""
    rng = _seeded_random(seed)
    return tuple(_draw_line(rng, _COIN_DISTRIBUTION) for _ in range(6))


def cast_yarrow_stalks(seed: str) -> tuple[LineKind, ...]:
    """Six-line cast using the yarrow-stalk probability table."""
    rng = _seeded_random(seed)
    return tuple(_draw_line(rng, _YARROW_DISTRIBUTION) for _ in range(6))


# ───── King Wen hexagram lookup ──────────────────────────────────────


# Bottom-up binary representation → King Wen number. Each line is 1
# for yang, 0 for yin; line 1 (bottom) is the least-significant bit
# in the string but the leftmost in our tuple ordering. We build a
# tuple-keyed lookup table for cleanest call-sites.
_KING_WEN_BINARY: Final[tuple[str, ...]] = (
    "111111",  # 1  Qián — the Creative
    "000000",  # 2  Kūn — the Receptive
    "100010",  # 3  Zhūn — Difficulty at the Beginning
    "010001",  # 4  Méng — Youthful Folly
    "111010",  # 5  Xū — Waiting
    "010111",  # 6  Sòng — Conflict
    "010000",  # 7  Shī — the Army
    "000010",  # 8  Bǐ — Holding Together
    "111011",  # 9  Xiǎo Chù — Small Taming
    "110111",  # 10 Lǚ — Treading
    "111000",  # 11 Tài — Peace
    "000111",  # 12 Pǐ — Standstill
    "101111",  # 13 Tóng Rén — Fellowship with Men
    "111101",  # 14 Dà Yǒu — Possession in Great Measure
    "001000",  # 15 Qiān — Modesty
    "000100",  # 16 Yù — Enthusiasm
    "100110",  # 17 Suí — Following
    "011001",  # 18 Gǔ — Work on the Decayed
    "110000",  # 19 Lín — Approach
    "000011",  # 20 Guān — Contemplation
    "100101",  # 21 Shì Hé — Biting Through
    "101001",  # 22 Bì — Grace
    "000001",  # 23 Bō — Splitting Apart
    "100000",  # 24 Fù — Return
    "100111",  # 25 Wú Wàng — Innocence
    "111001",  # 26 Dà Chù — Great Taming
    "100001",  # 27 Yí — the Corners of the Mouth
    "011110",  # 28 Dà Guò — Preponderance of the Great
    "010010",  # 29 Kǎn — the Abysmal (Water)
    "101101",  # 30 Lí — the Clinging (Fire)
    "001110",  # 31 Xián — Influence (Wooing)
    "011100",  # 32 Héng — Duration
    "001111",  # 33 Dùn — Retreat
    "111100",  # 34 Dà Zhuàng — Power of the Great
    "000101",  # 35 Jìn — Progress
    "101000",  # 36 Míng Yí — Darkening of the Light
    "101011",  # 37 Jiā Rén — the Family
    "110101",  # 38 Kuí — Opposition
    "001010",  # 39 Jiǎn — Obstruction
    "010100",  # 40 Jiě — Deliverance
    "110001",  # 41 Sǔn — Decrease
    "100011",  # 42 Yì — Increase
    "111110",  # 43 Guài — Breakthrough
    "011111",  # 44 Gòu — Coming to Meet
    "000110",  # 45 Cuì — Gathering Together
    "011000",  # 46 Shēng — Pushing Upward
    "010110",  # 47 Kùn — Oppression
    "011010",  # 48 Jǐng — the Well
    "101110",  # 49 Gé — Revolution
    "011101",  # 50 Dǐng — the Cauldron
    "100100",  # 51 Zhèn — the Arousing (Thunder)
    "001001",  # 52 Gèn — Keeping Still (Mountain)
    "001011",  # 53 Jiàn — Development
    "110100",  # 54 Guī Mèi — the Marrying Maiden
    "101100",  # 55 Fēng — Abundance
    "001101",  # 56 Lǚ — the Wanderer
    "011011",  # 57 Xùn — the Gentle (Wind)
    "110110",  # 58 Duì — the Joyous (Lake)
    "010011",  # 59 Huàn — Dispersion
    "110010",  # 60 Jié — Limitation
    "110011",  # 61 Zhōng Fú — Inner Truth
    "001100",  # 62 Xiǎo Guò — Preponderance of the Small
    "101010",  # 63 Jì Jì — After Completion
    "010101",  # 64 Wèi Jì — Before Completion
)


def _lines_to_binary(lines: tuple[bool, ...]) -> str:
    """Convert a bottom-up boolean tuple to the bit-string used by
    :data:`_KING_WEN_BINARY` (line 1 first)."""
    return "".join("1" if y else "0" for y in lines)


def _bool_lines(lines: tuple[LineKind, ...]) -> tuple[bool, ...]:
    return tuple(line.is_yang for line in lines)


def _bool_lines_after_change(lines: tuple[LineKind, ...]) -> tuple[bool, ...]:
    """Apply changing-line flips: old_yin → yang, old_yang → yin.
    Static lines keep their polarity."""
    out: list[bool] = []
    for line in lines:
        if line == LineKind.OLD_YIN:
            out.append(True)
        elif line == LineKind.OLD_YANG:
            out.append(False)
        else:
            out.append(line.is_yang)
    return tuple(out)


# Reverse lookup, populated at import time.
_BINARY_TO_NUMBER: Final[dict[str, int]] = {
    pattern: i + 1 for i, pattern in enumerate(_KING_WEN_BINARY)
}


def hexagram_for_lines(lines: tuple[bool, ...] | tuple[LineKind, ...]) -> int:
    """Return the King Wen number for a six-line pattern.

    Accepts either six booleans (True = yang) or six :class:`LineKind`
    values (interpreted by current polarity, *not* post-change). The
    bottom line comes first.
    """
    if len(lines) != 6:
        raise ValueError(f"expected 6 lines, got {len(lines)}")
    if all(isinstance(line, LineKind) for line in lines):
        bools = _bool_lines(lines)  # type: ignore[arg-type]
    else:
        bools = tuple(bool(line) for line in lines)
    return _BINARY_TO_NUMBER[_lines_to_binary(bools)]


def lines_for_hexagram(number: int) -> tuple[bool, ...]:
    """Inverse of :func:`hexagram_for_lines`. Returns the six-line
    bottom-up boolean pattern for King Wen ``number`` (1..64)."""
    if not 1 <= number <= 64:
        raise ValueError(f"King Wen number must be in [1, 64], got {number}")
    pattern = _KING_WEN_BINARY[number - 1]
    return tuple(c == "1" for c in pattern)


# ───── High-level cast ───────────────────────────────────────────────


def iching_cast(
    seed: str,
    method: CastMethod | str = CastMethod.THREE_COINS,
) -> CastResult:
    """Cast a reading deterministically.

    Returns the six lines, the primary hexagram number, the
    transformation hexagram number (if any lines change), and the
    1-based indices of changing lines.
    """
    method_enum = method if isinstance(method, CastMethod) else CastMethod(method)
    if method_enum == CastMethod.THREE_COINS:
        lines = cast_three_coins(seed)
    elif method_enum == CastMethod.YARROW_STALKS:
        lines = cast_yarrow_stalks(seed)
    else:  # pragma: no cover — enum exhaustive
        raise ValueError(f"unknown cast method: {method_enum!r}")

    primary = hexagram_for_lines(lines)

    changing_indices = tuple(
        i + 1 for i, line in enumerate(lines) if line.is_changing
    )
    if changing_indices:
        after = _bool_lines_after_change(lines)
        transformation: int | None = _BINARY_TO_NUMBER[_lines_to_binary(after)]
    else:
        transformation = None

    return CastResult(
        lines=lines,  # type: ignore[arg-type]
        primary_hexagram=primary,
        transformation_hexagram=transformation,
        changing_lines=changing_indices,
        method=method_enum,
    )

File: frontend/shared/src/divination/iching/engine.ts
/**
 * I Ching engine — port of the verbatim mockup engine from
 * `Theourgia I Ching.dc.html` (H04 handoff, lines 213-292).
 *
 * Two casting methods with deliberately different odds (H04 §S3.2):
 *   - Coin (3 coins): each coin heads/tails → sum + 6 ∈ {6,7,8,9},
 *     binomial distribution giving 6=1/8, 7=3/8, 8=3/8, 9=1/8.
 *   - Yarrow stalk: 6=1/16, 7=5/16, 8=7/16, 9=3/16. The yarrow rite
 *     is slower AND has different odds — the UI must respect both.
 *
 * Line values:
 *   6 = old yin   (changing, becomes yang)
 *   7 = young yang (stable)
 *   8 = young yin  (stable)
 *   9 = old yang  (changing, becomes yin)
 *
 * Hexagram resolution is purely derived from the six lines — never
 * stored. (See the H04 supplement §E worked example for the same
 * discipline applied to Geomancy.)
 */

// ─── Line values & predicates ─────────────────────────────────────

/** The four possible line values per the traditional rite. */
export type LineValue = 6 | 7 | 8 | 9;

/** True for the yang lines (7 = young yang, 9 = old yang). */
export function isYang(value: LineValue): boolean {
  return value === 7 || value === 9;
}

/** True for the changing lines (6 = old yin, 9 = old yang). */
export function isChanging(value: LineValue): boolean {
  return value === 6 || value === 9;
}

// ─── Trigrams ────────────────────────────────────────────────────

export interface Trigram {
  /** Pinyin name (Qian, Kun, …). */
  name: string;
  /** English image (Heaven, Earth, Thunder, …). */
  image: string;
  /** Unicode glyph (☰ ☷ ☳ ☵ ☱ ☶ ☲ ☴). */
  glyph: string;
}

/**
 * Trigrams indexed by their binary value:
 *   tv(line0, line1, line2) = yang(line0) + 2·yang(line1) + 4·yang(line2)
 *
 * Verbatim from `Theourgia I Ching.dc.html` lines 258 + 267. The
 * Pinyin name array (`trigName`) and the comp() array share this
 * index space.
 */
export const TRIGRAMS: readonly Trigram[] = [
  { name: "Kun", image: "Earth", glyph: "☷" }, // 0 = 000
  { name: "Zhen", image: "Thunder", glyph: "☳" }, // 1 = 100
  { name: "Kan", image: "Water", glyph: "☵" }, // 2 = 010
  { name: "Dui", image: "Lake", glyph: "☱" }, // 3 = 110
  { name: "Gen", image: "Mountain", glyph: "☶" }, // 4 = 001
  { name: "Li", image: "Fire", glyph: "☲" }, // 5 = 101
  { name: "Xun", image: "Wind", glyph: "☴" }, // 6 = 011
  { name: "Qian", image: "Heaven", glyph: "☰" }, // 7 = 111
];

/** Trigram look-up order used by the King-Wen table. Verbatim from
 *  `Theourgia I Ching.dc.html` line 259. */
const TRIGRAM_LOOKUP_ORDER: readonly string[] = [
  "Qian",
  "Zhen",
  "Kan",
  "Gen",
  "Kun",
  "Xun",
  "Li",
  "Dui",
];

/** The full 8×8 King-Wen lookup grid. Verbatim from line 260.
 *
 *   KING_WEN[lowerTrigramName][TRIGRAM_LOOKUP_ORDER.indexOf(upperTrigramName)]
 *
 * gives the King-Wen number ∈ 1..64. */
export const KING_WEN: Record<string, readonly number[]> = {
  Qian: [1, 34, 5, 26, 11, 9, 14, 43],
  Zhen: [25, 51, 3, 27, 24, 42, 21, 17],
  Kan: [6, 40, 29, 4, 7, 59, 64, 47],
  Gen: [33, 62, 39, 52, 15, 53, 56, 31],
  Kun: [12, 16, 8, 23, 2, 20, 35, 45],
  Xun: [44, 32, 48, 18, 46, 57, 50, 28],
  Li: [13, 55, 63, 22, 36, 37, 30, 49],
  Dui: [10, 54, 60, 41, 19, 61, 38, 58],
};

function trigramBinary(a: LineValue, b: LineValue, c: LineValue): number {
  return (isYang(a) ? 1 : 0) + (isYang(b) ? 2 : 0) + (isYang(c) ? 4 : 0);
}

/** The lower trigram (lines 1-3, indexes 0-2 in our bottom-up array). */
export function lowerTrigram(lines: readonly LineValue[]): Trigram {
  const v = trigramBinary(lines[0]!, lines[1]!, lines[2]!);
  return TRIGRAMS[v]!;
}

/** The upper trigram (lines 4-6, indexes 3-5). */
export function upperTrigram(lines: readonly LineValue[]): Trigram {
  const v = trigramBinary(lines[3]!, lines[4]!, lines[5]!);
  return TRIGRAMS[v]!;
}

/** Composition string in the mockup format:
 *  `"☰ Heaven over ☷ Earth"`. Verbatim from `comp()` line 271. */
export function trigramComposition(lines: readonly LineValue[]): string {
  const up = upperTrigram(lines);
  const lo = lowerTrigram(lines);
  return `${up.glyph} ${up.image} over ${lo.glyph} ${lo.image}`;
}

// ─── Hexagram resolution ─────────────────────────────────────────

/**
 * King-Wen number (1..64) for the six-line cast. Lines are bottom-up
 * (index 0 = bottom, index 5 = top), per the traditional rite.
 * Verbatim from `hexNum()` lines 257-265.
 */
export function hexagramNumber(lines: readonly LineValue[]): number {
  if (lines.length !== 6) {
    throw new Error(
      `hexagramNumber requires exactly 6 lines, got ${lines.length}`,
    );
  }
  const lo = lowerTrigram(lines).name;
  const up = upperTrigram(lines).name;
  const row = KING_WEN[lo];
  if (!row) {
    throw new Error(`No KING_WEN row for trigram ${lo}`);
  }
  const col = TRIGRAM_LOOKUP_ORDER.indexOf(up);
  if (col < 0) {
    throw new Error(`No KING_WEN column for trigram ${up}`);
  }
  return row[col]!;
}

export interface Transformation {
  /** 0-based indexes of the changing lines (6 or 9). Empty when the
   *  hexagram is stable. */
  changingLines: number[];
  /** King-Wen of the relating hexagram (after changing lines flip).
   *  Equal to the primary hexagram when no lines change. */
  relating: number;
}

/**
 * Compute the changing-line transformation. Verbatim from lines
 * 298-302: any 6 becomes 8 (yin→yang then becomes young yin?), no —
 * the mockup flips each changing line to its *opposite* fresh young
 * value: 9 → 8, 6 → 7. The relating hexagram is computed from the
 * post-flip lines.
 */
export function transformation(
  lines: readonly LineValue[],
): Transformation {
  const changingLines: number[] = [];
  const relLines: LineValue[] = lines.map((v, i) => {
    if (isChanging(v)) {
      changingLines.push(i);
      return v === 9 ? 8 : 7;
    }
    return v;
  });
  return {
    changingLines,
    relating: hexagramNumber(relLines),
  };
}

// ─── Hexagram names (CN + Pinyin + English) ──────────────────────

/**
 * Hexagram names indexed 1..64 (index 0 is the empty/unused slot from
 * the mockup arrays). Verbatim from `data()` lines 222-224. The full
 * Judgment + Image texts come from the backend; only the names ship
 * client-side.
 */
export const HEX_NAMES_CN: readonly string[] = [
  "",
  "乾",
  "坤",
  "屯",
  "蒙",
  "需",
  "訟",
  "師",
  "比",
  "小畜",
  "履",
  "泰",
  "否",
  "同人",
  "大有",
  "謙",
  "豫",
  "隨",
  "蠱",
  "臨",
  "觀",
  "噬嗑",
  "賁",
  "剝",
  "復",
  "無妄",
  "大畜",
  "頤",
  "大過",
  "坎",
  "離",
  "咸",
  "恆",
  "遯",
  "大壯",
  "晉",
  "明夷",
  "家人",
  "睽",
  "蹇",
  "解",
  "損",
  "益",
  "夬",
  "姤",
  "萃",
  "升",
  "困",
  "井",
  "革",
  "鼎",
  "震",
  "艮",
  "漸",
  "歸妹",
  "豐",
  "旅",
  "巽",
  "兌",
  "渙",
  "節",
  "中孚",
  "小過",
  "既濟",
  "未濟",
];

export const HEX_NAMES_PINYIN: readonly string[] = [
  "",
  "Qián",
  "Kūn",
  "Zhūn",
  "Méng",
  "Xū",
  "Sòng",
  "Shī",
  "Bǐ",
  "Xiǎo Chù",
  "Lǚ",
  "Tài",
  "Pǐ",
  "Tóng Rén",
  "Dà Yǒu",
  "Qiān",
  "Yù",
  "Suí",
  "Gǔ",
  "Lín",
  "Guān",
  "Shì Kè",
  "Bì",
  "Bō",
  "Fù",
  "Wú Wàng",
  "Dà Chù",
  "Yí",
  "Dà Guò",
  "Kǎn",
  "Lí",
  "Xián",
  "Héng",
  "Dùn",
  "Dà Zhuàng",
  "Jìn",
  "Míng Yí",
  "Jiā Rén",
  "Kuí",
  "Jiǎn",
  "Xiè",
  "Sǔn",
  "Yì",
  "Guài",
  "Gòu",
  "Cuì",
  "Shēng",
  "Kùn",
  "Jǐng",
  "Gé",
  "Dǐng",
  "Zhèn",
  "Gèn",
  "Jiàn",
  "Guī Mèi",
  "Fēng",
  "Lǚ",
  "Xùn",
  "Duì",
  "Huàn",
  "Jié",
  "Zhōng Fú",
  "Xiǎo Guò",
  "Jì Jì",
  "Wèi Jì",
];

export const HEX_NAMES_EN: readonly string[] = [
  "",
  "The Creative",
  "The Receptive",
  "Difficulty at the Beginning",
  "Youthful Folly",
  "Waiting",
  "Conflict",
  "The Army",
  "Holding Together",
  "Taming Power of the Small",
  "Treading",
  "Peace",
  "Standstill",
  "Fellowship with Men",
  "Great Possession",
  "Modesty",
  "Enthusiasm",
  "Following",
  "Work on the Decayed",
  "Approach",
  "Contemplation",
  "Biting Through",
  "Grace",
  "Splitting Apart",
  "Return",
  "Innocence",
  "Taming Power of the Great",
  "The Corners of the Mouth",
  "Great Preponderance",
  "The Abysmal Water",
  "The Clinging Fire",
  "Influence",
  "Duration",
  "Retreat",
  "Power of the Great",
  "Progress",
  "Darkening of the Light",
  "The Family",
  "Opposition",
  "Obstruction",
  "Deliverance",
  "Decrease",
  "Increase",
  "Breakthrough",
  "Coming to Meet",
  "Gathering Together",
  "Pushing Upward",
  "Oppression",
  "The Well",
  "Revolution",
  "The Cauldron",
  "The Arousing",
  "Keeping Still",
  "Development",
  "The Marrying Maiden",
  "Abundance",
  "The Wanderer",
  "The Gentle",
  "The Joyous",
  "Dispersion",
  "Limitation",
  "Inner Truth",
  "Small Preponderance",
  "After Completion",
  "Before Completion",
];

export interface HexagramName {
  number: number;
  chinese: string;
  pinyin: string;
  english: string;
}

export function hexagramName(num: number): HexagramName {
  if (num < 1 || num > 64) {
    throw new Error(`Invalid King-Wen number: ${num}`);
  }
  return {
    number: num,
    chinese: HEX_NAMES_CN[num]!,
    pinyin: HEX_NAMES_PINYIN[num]!,
    english: HEX_NAMES_EN[num]!,
  };
}

// ─── Cast ────────────────────────────────────────────────────────

export type IchingMethod = "coin" | "yarrow";

/**
 * Cast one line. Coin and yarrow give intentionally distinct odds.
 * Verbatim from `castOne()` lines 288-292.
 *
 *   Coin (3 coins flipped, sum + 6):
 *     6 = 1/8   (three tails)
 *     7 = 3/8   (one head)
 *     8 = 3/8   (two heads)
 *     9 = 1/8   (three heads)
 *
 *   Yarrow (the slower rite, distinct probabilities):
 *     6 = 1/16
 *     7 = 5/16
 *     8 = 7/16
 *     9 = 3/16
 */
export function castLine(
  method: IchingMethod,
  random: () => number = Math.random,
): LineValue {
  if (method === "coin") {
    let heads = 0;
    for (let k = 0; k < 3; k++) heads += random() < 0.5 ? 1 : 0;
    return (6 + heads) as LineValue;
  }
  // Yarrow odds — strictly different from coins.
  const r = random();
  if (r < 1 / 16) return 6;
  if (r < 6 / 16) return 7;
  if (r < 13 / 16) return 8;
  return 9;
}

/** Cast all six lines in one go (bottom-up order). The surface uses
 *  this for the coin-only "Cast all six" shortcut; the yarrow rite
 *  remains line-by-line to respect the meditative pacing. */
export function castSixLines(
  method: IchingMethod,
  random: () => number = Math.random,
): [LineValue, LineValue, LineValue, LineValue, LineValue, LineValue] {
  return [
    castLine(method, random),
    castLine(method, random),
    castLine(method, random),
    castLine(method, random),
    castLine(method, random),
    castLine(method, random),
  ];
}
